### PR TITLE
[REF] auth_brute_force: Cover all auth entrypoints

### PR DESCRIPTION
**Pull request by yajo**
To fix https://github.com/OCA/server-tools/issues/1125 I needed to refactor the addon.

The fix is affected by https://github.com/odoo/odoo/issues/24183 and will not work until it gets fixed upstream due to the technical limitations implied. ~~This PR is implemented assuming https://github.com/odoo/odoo/pull/24187 will be merged and backported.~~

@Tecnativa

### CODE REVIEW

**Comment by yajo** [comment]
According to https://github.com/odoo/odoo/pull/24187#issuecomment-381657198, I monkey-patched Odoo to get this safely working. As you might know, monkey patches affect all databases in the same instance, which can be bad in a SaaS environment. However, the patch is completely harmless in such case.

**Comment by pedrobaeza** [comment]
Using `_register_hook` y `_apply_patch` you do it only for the current DB.

**Comment by yajo** [comment]
Yikes, tricky tests! It's all :heavy_check_mark: now, ready to merge IMHO.

**Review by rafaelbn — APPROVED** [review]
Yes! Very important in terms of GDPR :+1: :

**Review by chienandalu — COMMENTED** [review]
With users behind the same IP (or bad configured proxies): if a user gets banned, would it put the other's sessions (already logged) down?

**Comment by yajo** [comment]
@chienandalu Those already logged in would retain their sessions, although new login would get locked, yes.

This is a side effect of this addon, docummented in Known Issues. Do you think there would be a better approach on the matter?

**Review by chienandalu — APPROVED** [review]
No, it's ok like that :)

**Comment by yajo** [comment]
I'm gonna think some more on the subject. It could be really problematic. Imagine:

- One office with 1 single public IP
- 200 users in that office
- At the same time, 5 of them fail passwords or users twice
- **Nobody else from the office will be able to log in**

So, the problem is quite big. I think it should be blocked just by user. Then maybe a notification could be sent to him.

File: auth_brute_force/models/res_authentication_attempt.py
Comment (LeartS) [inline auth_brute_force/models/res_authentication_attempt.py]: Maybe at least some of the fields used in this search should be setted as `index=True` ? On a medium/big installation running for years there could be potentially thousands (or even tens or hundreds of thousands) of login attempts to search in..

**Review by LeartS — APPROVED** [review]
Just code review, great code!

If someone tested this functionally, considering also the unittest, for me it's ok to merge.

**Comment by legalsylvain** [comment]
> So, the problem is quite big. I think it should be blocked just by user. Then maybe a notification could be sent to him.

I fully understand your use case. But if what you say is implemented, it should be optional. When I wrote this module, it was not designed to block user that doesn't remember their password, but to prevent attack and the only thing we have is the IP. (it's not perfect, it doesn't prevent Distributed attack, but it's a first step).

If it is only blocked by user and not by IP (as you suggest), it will generate the following use case : 

I want to attack a company that use odoo.
I know that the login is made with name@mycompany.com
I try to log in with all the login I know. (easy social attack).
Result : I can block all the login because my IP is not banned.

I see the implementation of the "block by login" as a big regression. 

**Solution 0 (no dev)**
- set a max_try to 30 (for exemple)
- improve password quality (to avoid attack by dictionnary)

-> I don't see any users trying to log in 30 times. After some tries, he will contact the administrator.


**solution 1**
- add a option "block by login" and / or "block by IP". The admin user can choose the policy, that depends on the network implementation of Odoo.

**solution 2**
- set hard_max_try to 10 (for exemple)
- add a soft_max_try to 5
make a improvment (by JS) to block user to attempt to login after 5 tries during 10 minutes, then during 1 hours, etc...

-> only one browser is blocked.
- maybe quite hard to develop

**solution 3**
- set hard_max_try to 10 (for exemple)
- add a soft_max_try to 5
make a improvment (by JS) to block users to log after 5 tries and redirect the user to a page to reinit the password, via a form. (classical token process).

what do you think ?

kind regards.

**Comment by LeartS** [comment]
Yes, that could be a problem. Adding the user on login attempts is in my opinion an ok solution; it would just mean that a potential attacker could try 10 passwords per user instead of 10 in general; which shouldn't be a problem.

File: auth_brute_force/models/res_authentication_attempt.py
Comment (legalsylvain) [reply auth_brute_force/models/res_authentication_attempt.py]: In the other hand, I don't see any interest to keep attempts with a date > X monthes. So maybe a cron to clean this table should be great. (I think there is a generic module that does that task for any table in server-tools.)

But Indeed, set an index=True is a good idea !

**Comment by LeartS** [comment]
> If it is only blocked by user and not by IP (as you suggest)

I think @Yajo meant to block just login attempts for that specific user to that IP, not to block the user in general. So the IP is still considered, and you can't use this to block access to odoo (unless you have the same IP as the real user, but in that case it's an internal problem and can't really be avoided..)

**Comment by yajo** [comment]
Yep, I think I'm gonna go that path: just block the user+IP combination instead of just the IP.

**Comment by legalsylvain** [comment]
> I think @Yajo meant to block just login attempts for that specific user to that IP, not to block the user in general. So the IP is still considered.

How ! I didn't understood ! Yes indeed. An option still is a good implementation. like 

```block_user_policy``` : 
* option 1 : block the IP ;
* option 2 : block the IP for the current login only.


> it would just mean that a potential attacker could try 10 passwords per user instead of 10 in general; which shouldn't be a problem.

It can be a problem. if you have a company with 1000 users, you let an attacker to make an attack with up to 10.000 attempts. he has more chance to succeed, statisticaly.

**Comment by LeartS** [comment]
> It can be a problem. if you have a company with 1000 users, you let an attacker to make an attack with up to 10.000 attempts. he has more chance to succeed, statisticaly.

Yes, 10000 in total, but still 10 per user. But I get what you mean: if an attacker tries the 10 most common passwords, the probability of at least 1 user in 1000 using it is much higher than the probability of a single specific user using it.
Maybe we could solve it this way, with two config values:

* Block remote-user after x invalid attempts (the param already implemented)
* Block remote generally after y invalid attempts

So, for example, block ip 1.1.1.1 for user 'admin' after 5 incorrect attempts for user 'admin', and block 1.1.1.1 in general after 50 consecutive invalid attempts between all users.
Setting x = y would mean to always block the IP, setting y to -1 could mean to never block the entire IP.

**Comment by LeartS** [comment]
@Yajo I was thinking it could be useful to save and show the login attempt channel (web - xmlrpc -jsonrpc). But it's ok to do it in a future PR as to not overload this one.

**Comment by yajo** [comment]
You'd be interested in v12 approach, currently in RD phase: https://github.com/odoo/odoo/pull/22212

**Comment by legalsylvain** [comment]
> So, for example, block ip 1.1.1.1 for user 'admin' after 5 incorrect attempts for user 'admin', and block 1.1.1.1 in general after 50 consecutive invalid attempts between all users.
Setting x = y would mean to always block the IP, setting y to -1 could mean to never block the entire IP.

Good idea ! very modular. :+1:

**Comment by yajo** [comment]
This has to change quite dramatically current implementation or it would be too hard to implement. Changing, though.

**Comment by yajo** [comment]
It changed now, take a look please!

**Comment by pedrobaeza** [comment]
Check Travis

**Comment by yajo** [comment]
It's weird, it seems some kinda conflict with other addon, i did a blind commit, let's see if it works.

**Comment by yajo** [comment]
OK Travis is :heavy_check_mark: now. I digged a lot and couldn't discover the conflict, so I just triggered separate test suites.

Shall we squash+merge now?

File: auth_brute_force/models/res_authentication_attempt.py
Comment (naglis) [inline auth_brute_force/models/res_authentication_attempt.py]: I think you can use `_logger.exception` instead of `_logger.error` with `exc_info=True`.

File: auth_brute_force/models/res_authentication_attempt.py
Comment (naglis) [inline auth_brute_force/models/res_authentication_attempt.py]: Nitpick: you may also use the [`.search_count()`](https://github.com/odoo/odoo/blob/14f59f4b780fa1e757cfce6a46c62c7a6b00d13e/odoo/models.py#L1486-L1494) shortcut.

File: auth_brute_force/models/res_users.py
Comment (naglis) [inline auth_brute_force/models/res_users.py]: Typo: *Remvove*

File: auth_brute_force/models/res_users.py
Comment (naglis) [inline auth_brute_force/models/res_users.py]: Nitpick: `{}` is [faster](https://doughellmann.com/blog/2012/11/12/the-performance-impact-of-using-dict-instead-of-in-cpython-2-7-2/) :trollface:

File: auth_brute_force/tests/test_brute_force.py
Comment (naglis) [inline auth_brute_force/tests/test_brute_force.py]: Typo: *Remvove*.

File: auth_brute_force/tests/test_brute_force.py
Comment (naglis) [inline auth_brute_force/tests/test_brute_force.py:74]: Nitpick: might be worth storing the `timeout` as a constant, seeing that it is used multiple times.

File: auth_brute_force/tests/test_brute_force.py
Comment (naglis) [inline auth_brute_force/tests/test_brute_force.py:77]: Nitpick / Consistency: here (and below) timeout is passed without the argument name where before the argument name is specified.

File: auth_brute_force/views/view.xml
Comment (naglis) [inline auth_brute_force/views/view.xml]: Nitpick: extra space before `/>`.

File: auth_brute_force/views/view.xml
Comment (naglis) [inline auth_brute_force/views/view.xml]: Nitpick: `colors` attribute [has been deprecated ](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#views) since v9.

File: auth_brute_force/models/res_authentication_attempt.py
Comment (naglis) [inline auth_brute_force/models/res_authentication_attempt.py]: Shouldn't it be `remote_metadata`?

**Comment by rafaelbn** [comment]
We need here the last approval of a PSC :+1:

File: auth_brute_force/README.rst
Comment (pedrobaeza) [inline auth_brute_force/README.rst:79]: Incorrect indentation

File: auth_brute_force/data/ir_config_parameter.xml
Comment (pedrobaeza) [inline auth_brute_force/data/ir_config_parameter.xml:9]: How do we control this now? No change in README?

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: auth_brute_force/models/res_authentication_attempt.py
Comment (naglis) [inline auth_brute_force/models/res_authentication_attempt.py]: Not sure about pros/cons of existing online service alternatives, but maybe it would be better to use a service which has HTTPS in their free plan, eg. https://ipapi.co/json?

Also, I have tested http://ip-api.com/json/ locally and after a couple of hundreds of requests I got banned (I guess timeout based), so in a real world case I believe this could potentially break if the limit is reached within the time window, resulting in empty metadata?

IIUC this would also be stored upon successful logins? In that case, each new `res.authentication.attempt` record could potentially take ~5 seconds? What if we try to fill the metadata asynchronously afterwards, eg. with Cron or connector? This way you could also group records by remote address and reduce the number of requests in this way  (now it is repeated for the same IP).

Also, I am wondering why an external service is used here instead of eg. [GeoIP](https://github.com/odoo/odoo/blob/14f59f4b780fa1e757cfce6a46c62c7a6b00d13e/addons/website/geoipresolver.py#L16)?

File: auth_brute_force/tests/test_brute_force.py
Comment (naglis) [inline auth_brute_force/tests/test_brute_force.py:115]: Readabillty: These test cases seem quite repetitive, maybe it would be possible to factor out the common parts to reduce code repetition?

File: auth_brute_force/models/res_authentication_attempt.py
Comment (naglis) [inline auth_brute_force/models/res_authentication_attempt.py:128]: Readability: The `# Check if remote is banned` and `# Check if remote + login combination is banned` parts share much of the code, might be worth factoring out into separate helper function.

**Review by naglis — APPROVED** [review]
@Yajo thank you so much for the effort, can't wait to test it out!

Added my 2¢, hopefully some of it will be helpful.

File: auth_brute_force/__manifest__.py
Comment (MiquelRForgeFlow) [inline auth_brute_force/__manifest__.py]: But this PR is in serve-tools 🤔

File: auth_brute_force/README.rst
Comment (MiquelRForgeFlow) [inline auth_brute_force/README.rst]: This issue was closed. Solved for v12

**Comment by rafaelbn** [comment]
@Yajo which is the state of this? OCA need this great work!

File: auth_brute_force/models/res_authentication_attempt.py
Comment (legalsylvain) [reply auth_brute_force/models/res_authentication_attempt.py]: Hi. I implemented the call of ip-api some years ago, but I did'nt made a benchmark about all the solutions, so maybe it's not the best solution.
Fill the metadata via a cron could be a great improvment indeed. And using GeoIP too ! (but maybe not the objective of this PR).

File: auth_brute_force/README.rst
Comment (yajo) [reply auth_brute_force/README.rst:79]: It's because it's a sub-point for previous point.

File: auth_brute_force/data/ir_config_parameter.xml
Comment (yajo) [reply auth_brute_force/data/ir_config_parameter.xml:9]: I updated the README.

File: auth_brute_force/models/res_authentication_attempt.py
Comment (yajo) [reply auth_brute_force/models/res_authentication_attempt.py]: I agree with both: this should be improved, but this PR targets other purpose. I added it to known issues.

File: auth_brute_force/models/res_authentication_attempt.py
Comment (yajo) [reply auth_brute_force/models/res_authentication_attempt.py]: I didn't want to give so much importance to this failure, which is not such a big problem. In fact, I think I'm gonna lower it to a warning instead :thinking:

File: auth_brute_force/tests/test_brute_force.py
Comment (yajo) [reply auth_brute_force/tests/test_brute_force.py:77]: Yes, that's because before we didn't have a 2nd arg.

File: auth_brute_force/tests/test_brute_force.py
Comment (yajo) [reply auth_brute_force/tests/test_brute_force.py:115]: Possibly, but the little details make it quite difficult to refactor, and after all repeating code in test cases is not so bad as in other parts of the code.

File: auth_brute_force/models/res_authentication_attempt.py
Comment (yajo) [reply auth_brute_force/models/res_authentication_attempt.py:128]: Done

**Comment by yajo** [comment]
Updated with all remarks attended.

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**
